### PR TITLE
Issue 429: Add component to input single file

### PR DESCRIPTION
**Pull request by jorgegonzalez**
### Summary of Changes

- Added UploadReport component to Upload page
	- Added 4 required UploadFile sections
- Added redux actions and reducers and modeled File Upload states
- Added error handling for uploads of invalid file types
- Added placeholder 'Submit Files' button for uploads

_Addresses issue_ #429
_Acceptance criteria as stated in the issue_

### How to Test

1. Log in as any user.
2. Click the Reports tab.
3. Select a year in the Fiscal Year dropdown.
4. Click Search.
5. Ensure a new UploadReport section appears below the ~~Begin Report~~ Search button with the four required sections on the page.
6. Drag a text file (`.txt` file extension) to any of the four sections.
7. Check that there were no errors, and the input area changes to display a file icon with the name of the uploaded file.
8. Try the same with a non-`.txt` file, and check that an error message is shown 
9. Clicking Submit Files should just refresh the page.
10. Clicking Cancel or changing the selected year should remove the UploadReport form.

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x] Files Input accepts a single file 
  - [x] Input accepts only -.txt file types
  - [x] Upload button is a placeholder
  - [x] Input displays a helpful error message when a wrong file type is uploaded 
  - [x] Frontend matches #427 delivered and accepted mockup
  - [x] Testing Checklist has been run and all tests pass

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: 99% (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [x] Did automated and manual testing with @reitermb  using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

Accessibility issues have been addressed

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces frontend code, is that code documented both inline and overall?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

No security issues detected

### CODE REVIEW

File: tdrs-frontend/package.json [hunk 1]
```diff
@@ -7,7 +7,6 @@
     "@fortawesome/free-solid-svg-icons": "^5.15.1",
     "@fortawesome/react-fontawesome": "^0.1.13",
     "axios": "^0.21.1",
-    "axios-cookiejar-support": "^1.0.1",
```
Comment (jorgegonzalez) [inline tdrs-frontend/package.json:10]: This was removed as it wasn't being used.

File: tdrs-frontend/src/assets/IdleTimer.scss [hunk 2]
```diff
@@ -1,12 +1,11 @@
 .modal {
   position: fixed; 
-  z-index: 1; 
+  z-index: 1000;
```
Comment (carltonsmith) [inline tdrs-frontend/src/assets/IdleTimer.scss:3]: @jorgegonzalez can you say something about why this change was made?

Comment (jorgegonzalez) [reply tdrs-frontend/src/assets/IdleTimer.scss:3]: It is not clear to me why this was needed as I didn't make this change. Reverting it locally, I didn't notice any differences in the IdleTimer modal. Should we reverse this?

Comment (carltonsmith) [reply tdrs-frontend/src/assets/IdleTimer.scss:3]: It may have been something related to accessibility, but I'm not sure since there are no comments. You might want to reach out to @reitermb to see if there was an issue surrounding this object

Comment (reitermb) [reply tdrs-frontend/src/assets/IdleTimer.scss:3]: I know we went through several iterations on the modal to check all the a11y boxes, but I'm not totally sure if the z-index here was one of those. 

If we change it I'd want to test again with a screenreader to make sure focus is still behaving as it should. E.g., Modal gets announced when it appears, focus remains inside it when tabbing around, focus goes back to where it was before the modal opened when it's closed.

My biggest question with this change would be whether it disrupts the focus going back to where it was bit.

Comment (carltonsmith) [reply tdrs-frontend/src/assets/IdleTimer.scss:3]: Ok - I think we should undo the change and redeploy the branch and check a11y. Then if we have to change it back we can add a comment indicating why the change was necessary

**Comment by jorgegonzalez** [comment]
Updated this PR based on the latest [TDP mockups](https://www.figma.com/file/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=3588%3A800); I will move this PR out of draft mode when I fix the tests.

**Comment by carltonsmith** [comment]
@jorgegonzalez Bringing this up locally I see that `Data Files` still shows up as `Reports`. As we want this to match the prototype I think we should make that change here.

**Comment by jorgegonzalez** [comment]
> @jorgegonzalez Bringing this up locally I see that Data Files still shows up as Reports. As we want this to match the prototype I think we should make that change here.

This should be addressed.

File: tdrs-frontend/src/components/Header/Header.test.js [hunk 3]
```diff
@@ -101,34 +101,34 @@ describe('Header', () => {
     expect(welcomeTab.hasClass('usa-current')).toEqual(true)
   })
 
-  it("should add usa-current class to Reports tab when on '/reports'", () => {
+  it("should add usa-current class to Reports tab when on '/data-files'", () => {
```
Comment (jtwillis92) [inline tdrs-frontend/src/components/Header/Header.test.js:104]:
```suggestion
  it("should add usa-current class to Data Files tab when on '/data-files'", () => {
```

File: tdrs-frontend/src/components/Header/Header.test.js [hunk 4]
```diff
@@ -45,9 +45,9 @@ describe('Header', () => {
         <Header />
       </Provider>
     )
-    const reportsLink = wrapper.find('#reports')
+    const reportsLink = wrapper.find('#data-files')
```
Comment (jtwillis92) [inline tdrs-frontend/src/components/Header/Header.test.js:48]:
```suggestion
    const dataFilesLink = wrapper.find('#data-files')
```

File: tdrs-frontend/src/components/Header/Header.test.js [hunk 5]
```diff
@@ -45,9 +45,9 @@ describe('Header', () => {
...
     expect(reportsLink).toExist()
-    expect(reportsLink).toIncludeText('Reports')
+    expect(reportsLink).toIncludeText('Data Files')
```
Comment (jtwillis92) [inline tdrs-frontend/src/components/Header/Header.test.js:50]:
```suggestion
    expect(dataFilesLink).toIncludeText('Data Files')
```
the line above this also should get changed if you take these suggestions.

File: tdrs-frontend/src/components/Header/Header.test.js [hunk 6]
```diff
@@ -101,34 +101,34 @@ describe('Header', () => {
...
+      router: { location: { pathname: '/data-files/upload' } },
     })
     const wrapper = mount(
       <Provider store={store}>
         <Header />
       </Provider>
     )
 
-    const reportsTab = wrapper.find('#reports')
+    const reportsTab = wrapper.find('#data-files')
```
Comment (jtwillis92) [inline tdrs-frontend/src/components/Header/Header.test.js:131]: A find and replace for reports => dataFiles might take care of these in this file.

**Review by jtwillis92 — COMMENTED** [review]
Looks good, left a couple small nits for consistency inside the same file.

File: tdrs-frontend/src/components/Header/Header.test.js (hunk 6 above)
Comment (jorgegonzalez) [reply tdrs-frontend/src/components/Header/Header.test.js]: Nice catch, thanks!

**Review by jtwillis92 — APPROVED** [review]

**Comment by jorgegonzalez** [comment]
Some screenshots.

**Review by reitermb — APPROVED** [review]

**Review by carltonsmith — APPROVED** [review]
LGTM!

File: tdrs-frontend/src/components/FileUpload/FileUpload.jsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        type="file"
+        name={sectionName}
+        aria-describedby={
+          hasUploadedFile
+            ? `${formattedSectionName}-file`
+            : formattedSectionName
+        }
+        aria-hidden="false"
+        accept=".txt"
+        data-errormessage="We can’t process that file format. Please provide a .txt file."
```
Comment (ADPennington) [inline tdrs-frontend/src/components/FileUpload/FileUpload.jsx:78]: @jorgegonzalez just curious: is there a way to capture what kind of file "that file format" is and reflect that to the user? e.g. 'We can't process <insert file extension> file formats. please provide a .txt file?'

File: tdrs-frontend/src/components/UploadReport/UploadReport.test.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+          fileName: null,
+          error: null,
+        },
+      ],
+    },
+  }
+  const mockStore = configureStore([thunk])
+  const handleCancel = jest.fn()
+
+  it('should render four inputs for uploading files', () => {
```
Comment (ADPennington) [inline tdrs-frontend/src/components/UploadReport/UploadReport.test.js:40]: just noting for the future, I might expect for the # of inputs in the initial state to be tied to the STT and based on whether sample data (section 4) is reported.

File: tdrs-frontend/src/actions/reports.test.js [hunk 9]
```diff
@@ -1,7 +1,28 @@
...
+
+    await store.dispatch(
+      upload({ file: { name: 'HELLO' }, section: 'Active Case Data' })
+    )
+
+    const actions = store.getActions()
+
+    expect(actions[0].type).toBe(SET_FILE)
+    expect(actions[0].payload).toStrictEqual({
+      fileName: 'HELLO',
```
Comment (ADPennington) [inline tdrs-frontend/src/actions/reports.test.js:21]: is there a reason the filename is 'hello'?

Comment (jorgegonzalez) [reply tdrs-frontend/src/actions/reports.test.js:21]: No; this is an arbitrary name in these redux action tests.

Comment (jorgegonzalez) [reply tdrs-frontend/src/actions/reports.test.js:21]: Note that this set of tests don't touch UI components so no file validation happens.

File: tdrs-frontend/src/reducers/reports.test.js [hunk 10]
```diff
@@ -1,9 +1,210 @@
...
+          fileName: null,
+          error: null,
+        },
+      ],
+      year: 2020,
+    })
+  })
+
+  it('should handle SET_FILE_ERROR', () => {
+    const fakeError = new Error({ message: 'something went wrong' })
```
Comment (ADPennington) [inline tdrs-frontend/src/reducers/reports.test.js:113]: @jorgegonzalez what does this mean and how might it occur?

File: tdrs-frontend/src/assets/uswds/_uswds-theme-utilities.scss [hunk 11]
```diff
@@ -31,7 +31,7 @@ used by utilities or layout grid
 */
 
 $theme-utility-breakpoints: (
-  "card": false,
+  "card": true,
```
Comment (ADPennington) [inline tdrs-frontend/src/assets/uswds/_uswds-theme-utilities.scss:34]: what's this change about? @jorgegonzalez

Comment (jorgegonzalez) [reply tdrs-frontend/src/assets/uswds/_uswds-theme-utilities.scss:34]: This turns on breakpoints for "card-sized" viewports (i.e. width of at least 160px).

See the `card` utility class is referenced in the Submit Files Button:

https://github.com/raft-tech/TANF-app/blob/0293259eadb6036348f636ac443aa0f7b69c59f6/tdrs-frontend/src/components/UploadReport/UploadReport.jsx#L43

For the smallest viewports, we need that class defined to provide additional margin to the button:

Before:
Screen Shot 2021-03-18 at 3 24 22 PM

After:
Screen Shot 2021-03-18 at 3 24 50 PM

**Comment by ADPennington** [comment]
@jorgegonzalez thanks so much for the helpful responses! below is feedback related to testing. QASP checklist forthcoming, pending your review of feedback below, a11y review, and demo! cc: @carltonsmith 

- [x]  Log in as any user.

*note related to [this issue](https://github.com/raft-tech/TANF-app/issues/239): I was logged in as an OFA admin, and didn't see any reference to the STT I'm affiliated with to upload files for. Also, the profile tab prompted me to request access to TDP which was counterintuitive. I'd like to see my profile information in this tab. (this has no impact on qasp review of this PR), but I'm wondering if we should revisit this for OFA MVP?*

- [x]  Click the Reports tab.

*can we add pictures of what we should see here? recommend considering this for all the testing steps.*

- [x]  Select a year in the Fiscal Year dropdown.
- [x]  Click Search.

*the code in places like [here](https://github.com/raft-tech/TANF-app/blob/5f0b2c2f3dbe9d9e58f66cf2638426ed65a01613/tdrs-frontend/src/components/Reports/Reports.jsx#L16) suggest that we might expect `begin report` buttons, which have been changed to `search`.* 

- [x]  Ensure a new UploadReport section appears below the Begin Report button with the four required sections on the page.

*see above note re: `begin report`*

- [x]  Drag a text file (`.txt` file extension) to any of the four sections.
- [x]  Check that there were no errors, and the input area changes to display a file icon with the name of the uploaded file.

*Noticed that dropping the same .txt file in multiple sections resulted in a spinning wheel that never ended. I was able to interrupt this by replacing with a non-dup .txt file.* 

- [x]  Try the same with a non-`.txt` file, and check that an error message is shown

*tried this successfully with: (1) folders, (2) pdfs, (3) xlsx and csv, and (4) dummy section 3 files with no extension. the latter is concerning because STTs will often not have an extension (e.g. ADS.E2J.FTP3.TS01) , so we'll want to accept those. Also tested that removing `accept '.txt` in browser dev tools allowed us to drag files w/o .txt extensions. which is expected behavior and a reminder of the importance of quarantining files until these are verified somehow.* 

- [x]  Clicking Submit Files should just refresh the page.

*this worked. noting here that refresh essentially meant starting over, which id expect (though id want to view files submitted to date—just thinking toward future. no impact on qasp).* 

- [x]  Clicking Cancel or changing the selected year should remove the UploadReport form.

**Comment by jorgegonzalez** [comment]
> Noticed that dropping the same .txt file in multiple sections resulted in a spinning wheel that never ended. I was able to interrupt this by replacing with a non-dup .txt file.

This appears to be a bug with the USWDS File Input component we're using. You can see it even happens on their [demo page](https://designsystem.digital.gov/components/file-input/).

**Comment by jorgegonzalez** [comment]
> tried this successfully with: (1) folders, (2) pdfs, (3) xlsx and csv, and (4) dummy section 3 files with no extension. the latter is concerning because STTs will often not have an extension (e.g. ADS.E2J.FTP3.TS01) , so we'll want to accept those.

So there is no way to limit `inputs` to accept `.txt` files and `no-extension` files. We're going to have to add our own file validation, something we [mentioned doing](https://github.com/raft-tech/TANF-app/pull/705#discussion_r597091660) earlier. 

Given that we have a [followup PR](https://github.com/raft-tech/TANF-app/pull/757) with incremental additions for #429, I'm thinking we can implement this change there so as to not bog down this PR furhter cc @carltonsmith

**Comment by jorgegonzalez** [comment]
> the code in places like here suggest that we might expect begin report buttons, which have been changed to search.

This has been fixed.

**Comment by carltonsmith** [comment]
> > tried this successfully with: (1) folders, (2) pdfs, (3) xlsx and csv, and (4) dummy section 3 files with no extension. the latter is concerning because STTs will often not have an extension (e.g. ADS.E2J.FTP3.TS01) , so we'll want to accept those.
> 
> So there is no way to limit `inputs` to accept `.txt` files and `no-extension` files. We're going to have to add our own file validation, something we [mentioned doing](https://github.com/raft-tech/TANF-app/pull/705#discussion_r597091660) earlier.
> 
> Given that we have a [followup PR](https://github.com/raft-tech/TANF-app/pull/757) with incremental additions for #429, I'm thinking we can implement this change there so as to not bog down this PR further cc @carltonsmith

The AC's for this issue explicitly state that we want to accept only `.txt` files, so I think it would make sense to create a new issue for it or add it to the ACs for the issue mentioned above or to 430, since that's where the actual upload happens.

**Comment by ADPennington** [comment]
> > > tried this successfully with: (1) folders, (2) pdfs, (3) xlsx and csv, and (4) dummy section 3 files with no extension. the latter is concerning because STTs will often not have an extension (e.g. ADS.E2J.FTP3.TS01) , so we'll want to accept those.
> > 
> > 
> > So there is no way to limit `inputs` to accept `.txt` files and `no-extension` files. We're going to have to add our own file validation, something we [mentioned doing](https://github.com/raft-tech/TANF-app/pull/705#discussion_r597091660) earlier.
> > Given that we have a [followup PR](https://github.com/raft-tech/TANF-app/pull/757) with incremental additions for #429, I'm thinking we can implement this change there so as to not bog down this PR further cc @carltonsmith
> 
> The AC's for this issue explicitly state that we want to accept only `.txt` files, so I think it would make sense to create a new issue for it or add it to the ACs for the issue mentioned above or to 430, since that's where the actual upload happens.

agreed during tech sync to add this as a priority to 430. cc: @lfrohlich

**Comment by ADPennington** [comment]
@jorgegonzalez -- I'll be interested to hear more about how the .cpp (c++) files were accepted. @carltonsmith also suggested that I try an html file in next round of testing. i'll report back here!

**Comment by ADPennington** [comment]
> > Noticed that dropping the same .txt file in multiple sections resulted in a spinning wheel that never ended. I was able to interrupt this by replacing with a non-dup .txt file.
> 
> This appears to be a bug with the USWDS File Input component we're using. You can see it even happens on their [demo page](https://designsystem.digital.gov/components/file-input/).
> 
> edit:
> 
> This isn't something we can fix without forking their component. Perhaps we should open an issue in the USWDS repo? I don't see this as an existing issue.

good to know! all agreed during drop-in that seems worth submitting an issue with USWDS, but also noting here for the record that we don't expect STTs to be submitting dup files across sections and wouldnt want to permit this 😄

**Comment by jorgegonzalez** [comment]
> The AC's for this issue explicitly state that we want to accept only .txt files, so I think it would make sense to create a new issue for it or add it to the ACs for the issue mentioned above or to 430, since that's where the actual upload happens.


#430 has been updated with this requirement.

**Comment by alexsoble** [comment]
Thanks very much for making that update @jorgegonzalez!

File: tdrs-frontend/src/components/FileUpload/FileUpload.jsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+            id={`${formattedSectionName}-error-alert`}
+            role="alert"
+          >
+            {file.error.message}
+          </div>
+        )}
+      </div>
+      <div
+        id={`${formattedSectionName}-file`}
+        aria-hidden
```
Comment (alexsoble) [inline tdrs-frontend/src/components/FileUpload/FileUpload.jsx:59]: Do we want the `ariaDescription` to have this `aria-hidden` class? 

Looks like `aria-hidden` will make the content unavailable, "removes that element and all of its children from the accessibility tree."

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-hidden_attribute

Comment (jorgegonzalez) [reply tdrs-frontend/src/components/FileUpload/FileUpload.jsx:59]: Yes this is deliberate so it is read for only screen readers.

Comment (jorgegonzalez) [reply tdrs-frontend/src/components/FileUpload/FileUpload.jsx:59]: https://www.accessibility-developer-guide.com/examples/sensible-aria-usage/describedby/

The `aria-describedby` attribute in the `input` below references this hidden div so that it is read by screen readers while hidden normally.

Comment (alexsoble) [reply tdrs-frontend/src/components/FileUpload/FileUpload.jsx:59]: Ah, thanks for the explanation! 😄

**Review by iamjolly — APPROVED** [review]
This is really helpful info. Thanks @jorgegonzalez

**Comment by iamjolly** [comment]
This looks great from an accessibility perspective. It's good to go!

**Review by lfrohlich — APPROVED** [review]
This looks really good! One thing I noticed, that I assume will be addressed in #430  is that I could add a JPG to the component through the dialog box.

**Comment by jorgegonzalez** [comment]
> This looks really good! One thing I noticed, that I assume will be addressed in #430 is that I could add a JPG to the component through the dialog box.

@lfrohlich that's correct. We added an AC to #430 to address the file type validation.
